cusub_cortex/state_machine/src/tasks/start_gate_det_adam.py

```diff
@@ -0,0 +1,411 @@
+            mag = np.sqrt(height * width)
+
+        if self.num_of_poles == 2:
+            [az_0, el_0, height_0, width_0] = self.listener.get_avg_bearing(self.dobj_nums[0], num_dv=5)
+            [az_1, el_1, height_1, width_1] = self.listener.get_avg_bearing(self.dobj_nums[1], num_dv=5)
+            
+            az = ((az_0 + az_1)/2)
+            el = ((el_0 + el_1)/2)
+            mag = np.sqrt(((height_0 * width_0) + (height_1 * width_1)) / 2)
+
+        if self.num_of_poles == 3:
+            [az, el, height, width] = self.listener.get_avg_bearing(self.pole_c, num_dv=5)
+            mag = np.sqrt(height * width)
+            
+        return [az, el, mag]
+    
+    def eval_startgate_poles(self):
+        [az_0, el_0, height_0, width_0] = self.listener.get_avg_bearing(self.dobj_nums[0], num_dv=5)
+        [az_1, el_1, height_1, width_1] = self.listener.get_avg_bearing(self.dobj_nums[1], num_dv=5)
+        [az_2, el_2, height_2, width_2] = self.listener.get_avg_bearing(self.dobj_nums[2], num_dv=5)
+
+        az_dict = {
+            float(az_0): self.dobj_nums[0],
+            float(az_1): self.dobj_nums[1],
+            float(az_2): self.dobj_nums[2]
+        }
+
+        az_output = sorted(az_dict, key = lambda x:float(x))
+
+        self.pole_l = self.dobj_nums[az_dict.get(az_output[2])]
+        self.pole_r = self.dobj_nums[az_dict.get(az_output[0])]
+        self.pole_c = self.dobj_nums[az_dict.get(az_output[1])]
+
+
+class Approach_2_Poles(Approach):
+    outcomes = ['found_3rd_pole', 'timed_out', 'lost_object']
+    num_of_poles = 2
+
+    def __init__(self, task_name, listener, clients):
+        super(Approach_2_Poles, self).__init__(task_name, listener, clients)
+
+    def execute(self, userdata):
+        ret = super(Approach_2_Poles, self).execute(userdata)
+        if ret == "found_pole":
+            return "found_3rd_pole"
+        else:
+            return ret
+
+
+class Approach_3_Poles(Approach):
+    outcomes = ['in_position', 'timed_out', 'lost_object']
+    num_of_poles = 3
+
+    def __init__(self, task_name, listener, clients):
+        super(Approach_3_Poles, self).__init__(task_name, listener, clients)
+
+    def execute(self, userdata):
+        ret = super(Approach_3_Poles, self).execute(userdata)
+        if ret == "found_pole":
+            return "in_position"
+        else:
+            return ret
+
+
+class Center_Orbit(Approach):
+    outcomes = ['centered', 'timed_out', 'lost_object']
+    num_of_poles = 3
+
+    def __init__(self, task_name, listener, clients):
+        super(Center_Orbit, self).__init__(task_name, listener, clients)
+
+    def execute(self, userdata):
+        ret = super(Center_Orbit, self).execute(userdata)
+        if ret == "found_pole":
+            return "centered"
+        else:
+            return ret
+
+    def action_func(self):
+        self.yaw_setpoint = self.az
+        if -0.03 < self.az < 0.03:
```

Comment:
self.az is the dobjects azimuth in the world frame from the subs position, not the orientation of the gate.
Potential termination conditions are ratio of azimuth between left/center and right/center pole or ratio of magintude of left and right poles.